### PR TITLE
=act #17530 fix BatchingExecutor vs. CallingThreadDispatcher

### LINKED ISSUE #17530: StackOverflowError awaiting future within akka
works with akka 2.3.10
fails with akka 2.3.11

our prototype's tests wait for some actors to be in place before actually running the tests (allowing smaller timeouts within the tests). it doesn't fail the first time but after some repetitive calls
The failing code called from within a actor's receive (! - we run tests in the same thread using sbt with fork=false, which seams to matter, as running the test from IDEA doesn't show the same errors):
```Await.result(system.actorSelection(system / myServiceActorName).resolveOne, serviceDuration)```
[info]   java.lang.StackOverflowError:
[info]   at java.lang.ThreadLocal$ThreadLocalMap.set(ThreadLocal.java:483)
[info]   at java.lang.ThreadLocal$ThreadLocalMap.access$100(ThreadLocal.java:298)
[info]   at java.lang.ThreadLocal.setInitialValue(ThreadLocal.java:184)
[info]   at java.lang.ThreadLocal.get(ThreadLocal.java:170)
[info]   at akka.dispatch.BatchingExecutor$AbstractBatch.resubmitUnbatched(BatchingExecutor.scala:60)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:104)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:106)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:106)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:106)
[info]   at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:106)

it might be related to #16327 introduced in 2.3.11

### PR DESCRIPTION
**Pull request by rkuhn**
fixes #17530

### CODE REVIEW

File: akka-actor-tests/src/test/scala/akka/dispatch/ExecutionContextSpec.scala [hunk 1]
```diff
@@ -106,6 +111,79 @@ class ExecutionContextSpec extends AkkaSpec with DefaultTimeout {
       }
       Await.result(f, 3.seconds) should be(())
     }
+
+    "work with tasks that block inside blocking" in {
+      system.dispatcher.isInstanceOf[BatchingExecutor] should be(true)
+      import system.dispatcher
+
+      val f = Future(()).flatMap { _ ⇒
+        blocking {
```
Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/dispatch/ExecutionContextSpec.scala]: blocking within blocking ought to be a noop, right?

Comment (rkuhn) [inline akka-actor-tests/src/test/scala/akka/dispatch/ExecutionContextSpec.scala]: yes, it ought to be—hence the test :-)

Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/dispatch/ExecutionContextSpec.scala]: Lol!

**Comment by patriknw** [comment]
LGTM

**Comment by drewhk** [comment]
LGTM

**Pull request merged**